typing/ikinds/ldd_hybrid_opt.ml

```diff
@@ -133,90 +147,115 @@ module Make (V : ORDERED) = struct
     then
-      let dh = leaf_value h in
-      let dl = down0 l in
-      if C.equal dl C.bot then h else leaf (C.co_sub dh dl)
-    else if is_leaf l
-    then canonicalize_right_leaf h l
+      let dh = Unsafe.leaf_value hi in
+      let dl = down0 lo in
+      if Axis_lattice.equal dl Axis_lattice.bot
+      then hi
+      else leaf (Axis_lattice.co_sub dh dl)
+    else if is_leaf lo
+    then canonicalize_right_leaf ~hi ~lo
     else
-      let hb = node_block h in
-      let lb = node_block l in
+      let hb = Unsafe.node_block hi in
+      let lb = Unsafe.node_block lo in
       let cmp = compare_var hb.v lb.v in
       if cmp = 0
       then
-        let lo' = canonicalize hb.lo lb.lo in
-        let hi1 = if lb.hi == bot then hb.hi else canonicalize hb.hi lb.hi in
-        let hi' = if lb.lo == bot then hi1 else canonicalize hi1 lb.lo in
-        if hb.lo == lo' && hb.hi == hi' then h else node_raw hb.v lo' hi'
+        let lo' = canonicalize ~hi:hb.lo ~lo:lb.lo in
+        let hi1 =
+          if lb.hi == bot
+          then hb.hi
+          else canonicalize ~hi:hb.hi ~lo:lb.hi
+        in
+        let hi' =
+          if lb.lo == bot
+          then hi1
+          else canonicalize ~hi:hi1 ~lo:lb.lo
+        in
+        if hb.lo == lo' && hb.hi == hi' then hi else node_raw hb.v lo' hi'
       else if cmp < 0
       then
-        let lo' = canonicalize hb.lo l in
-        let hi' = canonicalize hb.hi l in
-        if hb.lo == lo' && hb.hi == hi' then h else node_raw hb.v lo' hi'
-      else canonicalize h lb.lo
-
-  and canonicalize_right_leaf (h : node) (leaf_l : node) : node =
-    let leaf_val = leaf_value leaf_l in
-    if C.equal leaf_val C.bot then h else canonicalize_right_leaf_aux h leaf_val
-
-  and canonicalize_right_leaf_aux (h : node) (leaf_l : C.t) : node =
-    if is_leaf h
-    then leaf (C.co_sub (leaf_value h) leaf_l)
-    else
-      let hb = node_block h in
-      let lo' = canonicalize_right_leaf_aux hb.lo leaf_l in
-      let hi' = canonicalize_right_leaf_aux hb.hi leaf_l in
-      if hb.lo == lo' && hb.hi == hi' then h else node_raw hb.v lo' hi'
+        let lo' = canonicalize ~hi:hb.lo ~lo in
+        let hi' = canonicalize ~hi:hb.hi ~lo in
+        if hb.lo == lo' && hb.hi == hi' then hi else node_raw hb.v lo' hi'
+      else canonicalize ~hi ~lo:lb.lo
+
+  and canonicalize_right_leaf ~(hi : node) ~(lo : node) : node =
+    let rec aux ~(hi : node) (leaf_l : Axis_lattice.t) : node =
+      if is_leaf hi
+      then leaf (Axis_lattice.co_sub (Unsafe.leaf_value hi) leaf_l)
+      else
+        let hb = Unsafe.node_block hi in
+        let lo' = aux ~hi:hb.lo leaf_l in
+        let hi' = aux ~hi:hb.hi leaf_l in
+        if hb.lo == lo' && hb.hi == hi' then hi else node_raw hb.v lo' hi'
+    in
+    let leaf_val = Unsafe.leaf_value lo in
+    if Axis_lattice.equal leaf_val Axis_lattice.bot
+    then hi
+    else aux ~hi leaf_val
 
+  (* Build a canonical node; ensures [hi] is disjoint from [lo]. *)
```

Comment:
```suggestion
  (** Build a canonical node; ensures [hi] is disjoint from [lo]. *)
```
(My bar for whether a comment should get one or two stars is something like "would this comment be useful to appear when a user hovers over a usage of this identifier?")